clang/include/clang/AST/PreorderAST.h

```diff
@@ -46,6 +46,64 @@ namespace clang {
           assert(!isa<LeafExprNode>(Parent) &&
                  "Parent node cannot be a LeafExprNode");
       }
+
+    // Recursively coalesce BinaryOperatorNodes having the same commutative
+    // and associative operator.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Changed indicates whether a node was coalesced. We need this
+    // to control when to stop recursive coalescing.
+    // @param[in] Error indicates whether an error occurred during coalescing.
+    virtual void Coalesce(bool &Changed, bool &Error) { }
```

Comment:
It's good to make any function which does not have an implementation in the base class a pure virtual function meaning that its implementation is present only in the derived class.

For example:
`virtual void Coalesce(bool &Changed, bool &Error) = 0;`